### DISCUSSION TITLE
Can i use redirects in mutations?

### DISCUSSION THREAD

**Discussion by niklasgrewe** [blitz-js/blitz#2394] [Q&A]
Hi everyone,
i use a mutation for login. New users who have not yet set a username etc. should be redirected to another page. Can I do redirects only with `res.redirect` or in the `getServerSideProps` function but not in mutations? I would like to set redirects directly in the mutation function, like this:

```js
export default resolver.pipe(resolver.zod(Login), async ({ email }, ctx) => {
  const user = await authenticateUser(email)
  await ctx.session.$create({ userId: user.id, role: user.role as Role })
  
  if (!user.username) {
     return {
        redirect: {
           destination: "anotherpage"
           permament: false
        }
     }
  }
  return user
})
```

**Comment by flybayer** [comment]
Currently that's not possible, and it's unlikely we will change it. Queries and mutations are set up to not expose http by default.

In your case here, I suggest doing the redirect on the client right after login. Or possibly even adding the username to the session publicData and then have a declarative hook that redirects to a setup page if username is missing.

**Reply by niklasgrewe** [reply]
@flybayer ok, would the redirect on the client still work if I use `LoginPage.redirectAuthenticatedTo = "/"` on my login page? or would I then have to set the routing completely manually like this:

```js
const user = await loginMutation({ email: values.email })
if (!user.username) {
   router.push('/welcome')
} else {
   router.push('/dashboard')
}
```
> possibly even adding the username to the session publicData and then have a declarative hook that redirects to a setup page

sorry but what exactly do you mean by declarative hook, could you give me an example?

how can I secure the welcome page so that it can only be accessed by users who do not yet have a username, what would be best here? should I use the `userCurrentUser` hook here? but then I have to work with Suspense and Loading Spinner again, which would not be so advantageous for the UX.

**Reply by flybayer** [reply]
Yeah you can do like the code snippet you shared. Or in a hook like this:

```ts
const ensureOnboarded = () => {
  const session = useSession({suspense: false})
  useEffect(() => {
    if (!session.isLoading && !session.username) {
      router.push('/welcome')
    }
  }, [session]
}
```

And for the welcome page:

```ts
  const session = useSession({suspense: false})
  useEffect(() => {
    if (!session.isLoading && session.username) {
      router.push('/dashboard')
    }
  }, [session]
```

**Reply by niklasgrewe** [reply]
@flybayer thank you, but one thing I still do not quite understand... Where does the variable `user` come from in your example? to get the user i need to call the `useCurrentUser` hook, right? doesn't it perhaps make more sense to create a root context by storing the current user, so that you don't always have to use the `useCurrentUser` hook, which triggers a query every time?

**Reply by flybayer** [reply]
Oops! That should have been `session.username`, but yes it could also be `user` result from `useCurrentUser`. You need to modify your login/signup mutations to add username to the session publicData to access it this way.

It's ok to use the `useCurrentUser` hook as many places as you need. It internally, via useQuery, stores the data one time in Context

**Reply by niklasgrewe** [reply]
@flybayer ok thanks, it works with your provided code also good, but i noticed another some ux issue. When i manual go to the `/welcome` page, i can see the page content for a short time before i will be redirected. Is there a way to fix it, so i never seen the page content before i will be redirected? In this case, can i use `getServerSideProps()` or is there a better way do prevent it?

**Reply by flybayer** [reply]
If you don't want any flash, then yes `getServerSideProps` is your friend :)

**Reply by niklasgrewe** [reply]
@flybayer ok thanks, would you recommend me to do that in this case too? I read in the docs that caching is not so easy to set up here. Could the performance be affected?

**Reply by flybayer** [reply]
You don't have to worry about caching if you use `getServerSideProps` without `useQuery`. Caching only needed if you use both together.

For the UX you want, using `getServerSideProps` as often as you will make you the happiest :)

**Reply by niklasgrewe** [reply]
thank you, good to know 😊👍 but if I ever need `useQuery`, how can I do that with the caching then? I read online that some use [memory-cache](https://www.npmjs.com/package/memory-cache) etc.. Is something like this recommended? are there better methods?

**Reply by flybayer** [reply]
@niklasgrewe what caching are you talking about here? We have support to `prefetchQuery` inside gSSP which lets you use useQuery but without the initial loading flash: https://blitzjs.com/docs/query-usage#prefetching

**Reply by niklasgrewe** [reply]
@flybayer what do you mean with gSSP - getServerSideProps? What about when making db queries like this?

```js
export const getServerSideProps = async ({ req, res }) => {
  const session = await getSession(req, res)
  const result = await db.user.findUnique({
    where: { id: session.userId! },
    select: { email: true },
  })
  if (!session.username) {
    return {
      redirect: {
        destination: `/welcome?email=${result?.email}`,
        permanent: false,
      },
    }
  }
  return { props: {} }
}
```
can blitz also prefetch/cache this kind of queries? Is this approach recommend?

**Reply by MrLeebo** [reply]
If you are looking to cache database queries and other server operations like remote API calls, you're probably better off spinning up an instance of [redis](https://www.npmjs.com/package/redis) and caching your results there.

However, you probably wouldn't do that for something like this findUnique call, where the query is just seeking on an indexed column; the cost of accessing and writing the cache is probably more expensive than the DB call would be.

You'd want the cache for more complex reports (the kind of queries you would write in SQL directly instead of using the ORM)

Implementation-wise, using redis with blitz is not any different from how you'd use it in any other framework.

**Reply by flybayer** [reply]
@niklasgrewe that code snippet you have is a great way to handle it
